### PR TITLE
Legacy/port new imdb xpath changes

### PR DESCRIPTION
**Pull request by jsynowiec**
This merges upstream changes to parsers and fixes some issues. It allows to use legacy branch with recent IMDb website changes. Tested on a few tv shows and movies, seems to work.

Should be used with lxml because bs seems to be broken for some XPath queries. Don't know if it's the library itself or the imdbpy integration.

### CODE REVIEW

File: imdb/parser/http/movieParser.py [hunk 1]
```diff
@@ -164,276 +198,437 @@ class DOMHTMLMovieParser(DOMParserBase):
...
-                                        path="./img/@src"))
-                ]
+    extractors = [
+        Extractor(
+            label='title',
+            path="//meta[@property='og:title']",
+            attrs=Attribute(
+                key='title',
+                path="@content",
+                postprocess=analyze_og_title
```
Comment (drbig) [inline imdb/parser/http/movieParser.py]: Trailing comma.

File: imdb/parser/http/movieParser.py [hunk 2]
```diff
@@ -164,276 +198,437 @@ class DOMHTMLMovieParser(DOMParserBase):
...
+            )
```
Comment (drbig) [inline imdb/parser/http/movieParser.py]: Trailing comma.

File: imdb/parser/http/movieParser.py [hunk 3]
```diff
@@ -164,276 +198,437 @@ class DOMHTMLMovieParser(DOMParserBase):
...
+            group="//a[@class='glossary']",
+            group_key="./@name",
+            group_key_normalize=lambda x: x.replace('_', ' '),
+            path="../../../..//tr",
+            attrs=Attribute(
+                key=None,
+                multi=True,
+                path={
+                    'person': ".//text()",
+                    'link': "./td[1]/a[@href]/@href"
```
Comment (drbig) [inline imdb/parser/http/movieParser.py]: Trailing comma... Not sure I should point out every instance :)

File: imdb/parser/http/movieParser.py [hunk 4]
```diff
@@ -164,276 +198,437 @@ class DOMHTMLMovieParser(DOMParserBase):
...
+                },
+                postprocess=lambda x: build_person(
+                    x.get('person') or '',
```
Comment (drbig) [inline imdb/parser/http/movieParser.py]: Unless `x` is something non-dict conforming: `x.get('person', '')`.

File: imdb/parser/http/movieParser.py [hunk 5]
```diff
@@ -164,276 +198,437 @@ class DOMHTMLMovieParser(DOMParserBase):
...
+            attrs=Attribute(
+                key="cast",
+                multi=True,
+                path={
+                    'person': ".//text()",
+                    'link': "td[2]/a/@href",
+                    'roleID': "td[4]//div[@class='_imdbpyrole']/@roleid"
+                },
+                postprocess=lambda x: build_person(
+                    x.get('person') or '',
```
Comment (drbig) [inline imdb/parser/http/movieParser.py]: Ditto.

File: imdb/parser/http/movieParser.py [hunk 6]
```diff
@@ -444,15 +639,16 @@ def preprocess_dom(self, dom):
...
             tn15more.drop_tree()
         return dom
 
     re_space = re.compile(r'\s+')
     re_airdate = re.compile(r'(.*)\s*\(season (\d+), episode (\d+)\)', re.I)
+
     def postprocess_data(self, data):
         # Convert section names.
-        for sect in data.keys():
+        for sect in list(data.keys()):
```
Comment (Nozdi) [inline imdb/parser/http/movieParser.py]: In python 2 `keys()` is always a list. Also if we are just iterating we can do `for sect in data:`

File: imdb/parser/http/movieParser.py [hunk 7]
```diff
@@ -483,10 +679,8 @@ def postprocess_data(self, data):
                 del data['other akas']
             if nakas:
                 data['akas'] = nakas
-        if 'color info' in data:
```
Comment (Nozdi) [inline imdb/parser/http/movieParser.py]: completely removing?

File: imdb/parser/http/movieParser.py [hunk 8]
```diff
@@ -498,18 +692,22 @@ def postprocess_data(self, data):
...
+                    pass
                 if date and date != '????':
                     data['original air date'] = date
                 else:
                     del data['original air date']
                 # Handle also "episode 0".
-                if season or type(season) is type(0):
+                if season or isinstance(season, int):
                     data['season'] = season
-                if episode or type(season) is type(0):
```
Comment (Nozdi) [inline imdb/parser/http/movieParser.py]: :+1:

File: imdb/parser/http/movieParser.py [hunk 9]
```diff
@@ -683,41 +905,41 @@ def preprocess_dom(self, dom):
...
-                #      "insert" method.
                 tr.insert(position, clone)
         return dom
 
     def postprocess_data(self, data):
         if len(data) == 0:
             return {}
         nd = []
-        for key in data.keys():
+        for key in list(data.keys()):
```
Comment (Nozdi) [inline imdb/parser/http/movieParser.py]: ditto

File: imdb/parser/http/movieParser.py [hunk 10]
```diff
@@ -801,12 +1041,18 @@ class DOMHTMLTriviaParser(DOMParserBase):
...
+
+    extractors = [
+        Extractor(
+            label='alternate versions',
+            path="//div[@class='sodatext']",
+            attrs=Attribute(
+                key='trivia',
+                multi=True,
+                path=".//text()",
+                postprocess=lambda x: x.strip())
```
Comment (Nozdi) [inline imdb/parser/http/movieParser.py]: how about closing parenthesis in new line?

File: imdb/parser/http/movieParser.py [hunk 11]
```diff
@@ -1041,88 +1341,107 @@ class DOMHTMLRatingsParser(DOMParserBase):
         rparser = DOMHTMLRatingsParser()
         result = rparser.parse(userratings_html_string)
     """
-    re_means = re.compile('mean\s*=\s*([0-9]\.[0-9])\.\s*median\s*=\s*([0-9])',
```
Comment (Nozdi) [inline imdb/parser/http/movieParser.py]: Are we removing this dot on purpose?

File: imdb/parser/http/movieParser.py [hunk 12]
```diff
@@ -1863,24 +2433,25 @@ class DOMHTMLAiringParser(DOMParserBase):
...
+        )
     ]
 
     def postprocess_data(self, data):
         if len(data) == 0:
             return {}
-        seriesTitle = data['series title']
-        seriesID = analyze_imdbid(data['series id'])
-        if data.has_key('airing'):
+        seriesTitle = data.get('series title') or ''
```
Comment (Nozdi) [inline imdb/parser/http/movieParser.py]: `data.get('series title', '')`

File: imdb/parser/http/movieParser.py (hunk 7 above)
Comment (jsynowiec) [reply imdb/parser/http/movieParser.py]: Yes, it's moved outside this extractor, as a "standalone"

**Pull request merged**